datafusion/src/row/reader.rs

```diff
@@ -0,0 +1,441 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Accessing row from raw bytes
```

Comment:
I think a pretty picture would be very helpful, showing how data is encoded

Reply:
Here is one for consieration:

```text
                                                                                                        
 Row Layout                                                                                             
                                                                                                        
┌────────────────┬──────────────────────────┬───────────────────────┐        ┌───────────────────────┐  
│Validity Bitmask│    Fixed Width Field     │ Variable Width Field  │   ...  │     vardata area      │  
│ (byte aligned) │   (native type width)    │(len + vardata offset) │        │   (variable length)   │  
└────────────────┴──────────────────────────┴───────────────────────┘        └───────────────────────┘  
                                                                                                        
                                                                                                        
                                                                                                        
 For example, given the schema (Int8, Float32, Utf8, Utf8)                                              
                                                                                                        
 Encoding the tuple (1, NULL, "FooBar", "baz")                                                          
                                                                                                        
 Requires 35 bytes as shown                                                                             
┌────────┬────────┬──────────────┬──────────────────────┬──────────────────────┬───────────────────────┐
│0b000110│  0x01  │  0x00000000  │0x00000000  0x00000006│0x00000006  0x00000003│       FooBarbaz       │
└────────┴────────┴──────────────┴──────────────────────┴──────────────────────┴───────────────────────┘
0        1         2             10                     18                     26                     35
                                                                                                        
 Validity    Int8  Float32 Field       Utf8 Field 1         Utf8 Field 2          Variable length       
   Mask     Field    (4 bytes)           Offset: 0            Offset: 6                area             
 (1 byte)  (1 byte)                       Size: 6              Size: 3               (9 bytes)          
                                         (8 bytes)            (8 bytes)                                 
```

Also attaching the monopic file in case anyone finds that useful: 
[drawing.zip](https://github.com/apache/arrow-datafusion/files/8041744/drawing.zip)

Reply:
I've adapted the fig a little bit and put it on the module doc. Many thanks, @alamb!

Reply:
> The diagram is excellent! I put a to-do item yesterday but haven't got a chance to finish it yet. Do you mind I put this into the doc?

I don't mind at all -- that is why I made it :)